packages/katello/katello/hostname-change.rb

```diff
@@ -239,6 +240,135 @@ def setup_opt_parser
+      # find the character A enclosed by any whitespace chars
+      a_record = zone_file_data.find { |line| !(line =~ /\sA\s/).nil? }
+      ip = a_record.split('A').last.strip if a_record
+      serial_line = zone_file_data.find { |line| line.downcase.include?('serial') }
+      serial = serial_line.split('Serial').first.to_i if serial_line
+      {
+        ip: ip,
+        serial: serial
+      }
+    end
+
+    def new_dns_values
+      STDOUT.puts 'gathering DNS data...'
+      local_ip = '127.0.0.1'
+      zone = @scenario_answers['foreman_proxy']['dns_zone']
+      reverse_zone = @scenario_answers['foreman_proxy']['dns_reverse']
+      if reverse_zone.is_a?(Array)
+        reverse_zone = reverse_zone.first
+      end
+      soa_admin_domain = "root.#{zone}"
+      key_file = @scenario_answers['foreman_proxy']['keyfile']
+      begin
+        STDOUT.puts 'querying local DNS server...'
+        ip_serial_data = query_dns_server(@old_hostname, zone, local_ip)
+        ip = ip_serial_data[:ip]
+        serial = ip_serial_data[:serial]
+        new_serial = (serial + 1).to_s if serial
+        reverse_serial = query_dns_server(@old_hostname, reverse_zone, local_ip)[:serial]
+        new_reverse_serial = (reverse_serial + 1).to_s if reverse_serial
+      rescue Resolv::ResolvError => e
+        STDOUT.puts e
+        STDOUT.puts 'Parsing zone file as fallback'
+        ip_serial_data = parse_zone_file("/var/named/dynamic/db.#{zone}")
+        ip = ip_serial_data[:ip]
+        serial = ip_serial_data[:serial]
+        new_serial = (serial + 1).to_s if serial
+        reverse_serial = parse_zone_file("/var/named/dynamic/db.#{reverse_zone}")[:serial]
+        new_reverse_serial = (reverse_serial + 1).to_s if reverse_serial
+      ensure
+        run_cmd('rndc thaw')
+      end
+
+
+      {
+        local_ip: local_ip,
+        zone: zone,
+        old_fqdn: @old_hostname,
+        soa_admin_domain: soa_admin_domain,
+        ip: ip,
+        new_serial: new_serial,
+        new_reverse_serial: new_reverse_serial,
+        new_fqdn: @new_hostname,
+        key_file: key_file,
+        reverse_zone: reverse_zone
+      }
+    end
+
+    def nsupdate_command(dns_entries, key_file)
+      "echo -e \"#{dns_entries}\" | nsupdate -l -k #{key_file}"
+    end
+
+    def update_dns_records(new_dns_values = {})
+      new_dns_values.each do |key, value|
+        unless value
+          raise "Error gathering DNS data: couldn't find value for '#{key}'"
+        end
+      end
+
+      local_ip = new_dns_values[:local_ip]
+      zone = new_dns_values[:zone]
+      old_fqdn = new_dns_values[:old_fqdn]
+      soa_admin_domain = new_dns_values[:soa_admin_domain]
+      ip = new_dns_values[:ip]
+      new_serial = new_dns_values[:new_serial]
+      new_reverse_serial = new_dns_values[:new_reverse_serial]
+      new_fqdn = new_dns_values[:new_fqdn]
+      key_file = new_dns_values[:key_file]
+      reverse_zone = new_dns_values[:reverse_zone]
```

Comment:
@jeremylenz I'm having a hard time coming to terms with these lines, while its valid, it feels like a lot of work to get information from an already structured hash.  Are there any other alternatives?

How about using an [openstruct](https://docs.ruby-lang.org/en/2.0.0/OpenStruct.html) if you are concerned about the amount of characters in [my other suggestion](https://github.com/theforeman/foreman-packaging/pull/4706#discussion_r378980457)?
```ruby
d = OpenStruct.new(new_dns_values)
## then refer to the struct
d.key_file
d.new_fqdn
```
ideally `d` is as locally scoped as possible, breaking into individual methods like my first example. Open to other ideas here though, that is just first the one comes to mind that works with ruby 2.0.

Reply:
@johnpmitsch this looks like a good solution!  Updated.